dipy/align/mattes.pyx

```diff
@@ -0,0 +1,1390 @@
+            _joint_pdf_gradient_sparse_3d(theta, transform, sval, mval,
+                                          sample_points, mgradient, self.smin,
+                                          self.sdelta, self.mmin, self.mdelta,
+                                          self.nbins, self.padding,
+                                          self.joint_grad)
+        else:
+            msg = 'Only dimensions 2 and 3 are supported. '+str(dim)+' received'
+            raise ValueError(msg)
+
+
+    def update_mi_metric(self, update_gradient=True):
+        r""" Computes current value and gradient (if requested) of the MI metric
+
+        Parameters
+        ----------
+        update_gradient : Boolean
+            boolean indicating if the gradient must be computed (if False,
+            only the value is computed). Default is True.
+        """
+        grad = None
+        if update_gradient:
+            sh = self.joint_grad.shape[2]
+            if (self.metric_grad is None) or (self.metric_grad.shape[0] != sh):
+                self.metric_grad = np.empty(sh)
+            grad = self.metric_grad
+
+        self.metric_val = _compute_mattes_mi(self.joint, self.joint_grad,
+                                             self.smarginal, self.mmarginal,
+                                             grad)
+
+
+cdef inline double _bin_normalize(double x, double mval, double delta) nogil:
+    r''' Normalizes intensity x to the range covered by the Parzen histogram
+    We assume that mval was computed as:
+
+    (1) mval = xmin / delta - padding
+
+    where xmin is the minimum observed image intensity and delta is the
+    bin size, computed as:
+
+    (2) delta = (xmax - xmin)/(nbins - 2 * padding)
+
+    If the minimum and maximum intensities were assigned to the first and last
+    bins (with no padding), it could be possible that samples at the first and
+    last bins contribute to "non-existing" bins beyond the boundary (because
+    the support of the Parzen window may be larger than one bin). The padding
+    bins are used to collect such contributions (i.e. the probability of
+    observing a value beyond the minimum and maximum observed intensities may
+    correctly be assigned a positive value).
+
+    The normalized intensity is (from eq(1) ):
+
+    (3) nx = (x - xmin) / delta + padding = x / delta - mval
+
+    This means that normalized intensity nx must lie in the closed interval
+    [padding, nbins-padding], which contains bins with indices
+    padding, padding+1, ..., nbins - 1 - padding (i.e., nbins - 2*padding bins)
+
+    '''
+    return x / delta - mval
+
+
+cdef inline cnp.npy_intp _bin_index(double normalized, int nbins,
+                                    int padding) nogil:
+    r''' Index of the bin in which the normalized intensity 'normalized' lies
+    The intensity is assumed to have been normalized to the range of intensities
+    covered by the histogram: the bin index is the integer part of the argument,
```

Comment:
What do you mean by 'the integer part of the argument'?  Sorry if I am being slow.